shamil_latypov/main.cpp

```diff
@@ -1,138 +1,248 @@
+    return depth_map_[depth].size();
+  }
+
   // Добавляет ребро в graph
   void add_edge(const VertexId& v1, const VertexId& v2) {
     assert(has_vertex_id(v1) && "Vertex 1 doesnt exist\n");
     assert(has_vertex_id(v2) && "Vertex 2 doesnt exist\n");
-    assert(vertices_connected(v1, v2) && "Vertices are connected\n");
-    const auto& new_edge = edges_.emplace_back(v1, v2, get_new_edge_id());
-    vertices_[v1].add_edge_id(new_edge.get_id());
-    vertices_[v2].add_edge_id(new_edge.get_id());
+    assert(!vertices_connected(v1, v2) && "Vertices are connected\n");
+
+    int color_param = 0;
+    // По логике нашей задачи, если вторая вершина не имеет присоединенных к ней
+    // рёбер, значит добавляемое ребро серое
+    if (get_vertex(v2).get_edge_ids().size() == 0) {
+      // vertex2.depth = vertex1.depth + 1
+      get_vertex(v2).set_depth(get_vertex(v1).get_depth() + 1);
+
+      if (depth_map_.size() <= get_vertex(v2).get_depth()) {
+        depth_map_.push_back({});
+      }
+
+      // Добавление на нужную глубину vertex2.id и удаление vertex2.id из
+      // нулевой глубины
+      depth_map_[get_vertex(v2).get_depth()].push_back(v2);
+      for (auto vertex_id_it = depth_map_[0].begin();
+           vertex_id_it != depth_map_[0].end(); vertex_id_it++) {
+        if (*vertex_id_it == v2) {
+          depth_map_[0].erase(vertex_id_it);
+          break;
+        }
+      }
+    } else {  // Иначе задаем цвет ребру
+      color_param = set_edge_color(v1, v2);
+    }
+    const auto& new_edge =
+        edges_.emplace_back(v1, v2, get_new_edge_id(), color_param);
+    get_vertex(v1).add_edge_id(new_edge.get_id());
+    if (v1 != v2) {  // Чтобы id зеленых рёбер не повторялись
+      get_vertex(v2).add_edge_id(new_edge.get_id());
+    }
   }
 
   // Добавляет вершину в граф
-  void add_vertex() { vertices_.emplace_back(get_new_vertex_id()); }
+  void add_vertex() {
+    if (depth_map_.size() == 0) {
+      depth_map_.push_back({});
+    }
+    Vertex new_vertex = vertices_.emplace_back(get_new_vertex_id());
+    depth_map_[0].push_back(new_vertex.get_id());
+  }
 
-  // Возврат векторов
+  // Вектор вершин
   const std::vector<Vertex>& get_vertices() const { return vertices_; }
+  // Вектор рёбер
   const std::vector<Edge>& get_edges() const { return edges_; }
+  // Вектор карты глубины
+  const std::vector<std::vector<VertexId>>& get_depth_map() const {
+    return depth_map_;
+  }
 
-  bool vertices_connected(const VertexId& v1, const VertexId& v2) const {
-    for (const auto& edge_id1 : vertices_[v1].get_edge_ids()) {
-      for (const auto& edge_id2 : vertices_[v2].get_edge_ids()) {
-        if (edge_id1 == edge_id2) {
-          return false;
-        }
+  // Возврат вершины
+  const Vertex& get_vertex(const VertexId& id) const {
+    for (auto& vertex : vertices_) {
+      if (vertex.get_id() == id) {
+        return vertex;
       }
     }
-    return true;
+    return vertices_.back();
```

Comment:
`back()` - это последний элемент. В корне неверно по-умолчанию возвращать послений!
1) Добавь проверку входных параметров, что вершина существует
2) В таком случае у нас будет ожидание того, что вершина всегда должна вернуться, а значит можно просто выкинуть `exception`
```suggestion
    throw std::runtime_error("Vertex doesn't exist");
```